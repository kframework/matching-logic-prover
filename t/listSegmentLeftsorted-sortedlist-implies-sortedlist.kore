imports "t/definitions/listSegmentLeftSorted.kore"
imports "t/definitions/listSorted.kore"

/* lseglsorted(H, X, Y, F, PV, MAX) ==
     (X = Y and F = empty and PV <= MAX) OR
     (exists X', F', V. (H[X] = X' and F' U {X} = F and F' disjoint {X}) and
                        H[X+1] = V and X > 0 and V > PV and V <= MAX and
                        lseglsorted(H, X', Y, F', V, MAX))
*/

/* listsorted(H, X, F, PV) ==
     (X = 0 and F = empty) OR
     (exists X', F', V. (H[X] = X' and F' U {X} = F and F' disjoint {X}) and
                        H[X+1] = V and X > 0 and V > PV and
                        listsorted(H, X', F', V))
*/

claim \implies( \and( listSegmentLeftSorted( H { ArrayIntInt } 
                                    , X { Int } 
                                    , Y { Int } 
                                    , F { SetInt }
                                    , MIN { Int }
                                    , MAX { Int } 
                                    )
                    , listSorted( H { ArrayIntInt }
                                , Y { Int }
                                , G { SetInt }
                                , MIN2 { Int }) 
                    , \equals( K { SetInt }, union(F { SetInt }, G { SetInt })) 
                    , disjoint(F { SetInt }, G { SetInt }) 
                    , \not(gt(MAX { Int }, MIN2 { Int }))
                     )
              , \and( listSorted( H { ArrayIntInt }
                                , X { Int }
                                , K { SetInt }
                                , MIN { Int }
                                )
                     )
              )

strategy kt . normalize
            . ( ( simplify
                . left-unfold-Nth(0) .
                  ( ( right-unfold-Nth(0, 0) . smt)
                  | ( right-unfold-Nth(0, 1) . smt)
                  )
                )
              )

// strategy search-bound(4)
