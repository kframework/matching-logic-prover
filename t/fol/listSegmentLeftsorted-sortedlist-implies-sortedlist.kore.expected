<prover>
  <k>
    .
  </k>
  <exit-code>
    0
  </exit-code>
  <goals>
    <goal>
      active: true, id: root, parent: .
      <k>
        \implies ( \and ( listSegmentLeftSorted ( H { ArrayIntInt } , X { Int } , Y { Int } , F { SetInt } , MIN { Int } , MAX { Int } , .Patterns ) , listSorted ( H { ArrayIntInt } , Y { Int } , G { SetInt } , MIN2 { Int } , .Patterns ) , \equals ( K { SetInt } , union ( F { SetInt } , G { SetInt } , .Patterns ) ) , disjoint ( F { SetInt } , G { SetInt } , .Patterns ) , \not ( gt ( MAX { Int } , MIN2 { Int } , .Patterns ) ) , .Patterns ) , \and ( listSorted ( H { ArrayIntInt } , X { Int } , K { SetInt } , MIN { Int } , .Patterns ) , .Patterns ) )
      </k>
      <strategy>
        success
      </strategy>
      <expected>
        success
      </expected>
    <trace>
      .
    </trace>
  </goals>
  <declarations>
    <declaration>
      axiom \forall { H { ArrayIntInt } , X { Int } , F { SetInt } , PREV_VAL { Int } , .Patterns } \iff-lfp ( listSorted ( H { ArrayIntInt } , X { Int } , F { SetInt } , PREV_VAL { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( X { Int } , 0 ) , \equals ( F { SetInt } , emptyset ) , .Patterns ) , \exists { F_I { SetInt } , VAL_I { Int } , X_I { Int } , .Patterns } \and ( listSorted ( H { ArrayIntInt } , X_I { Int } , F_I { SetInt } , VAL_I { Int } , .Patterns ) , gt ( X { Int } , 0 , .Patterns ) , \equals ( select ( H { ArrayIntInt } , X { Int } , .Patterns ) , X_I { Int } ) , \equals ( F { SetInt } , union ( F_I { SetInt } , singleton ( X { Int } , .Patterns ) , .Patterns ) ) , disjoint ( F_I { SetInt } , singleton ( X { Int } , .Patterns ) , .Patterns ) , \equals ( VAL_I { Int } , select ( H { ArrayIntInt } , plus ( X { Int } , 1 , .Patterns ) , .Patterns ) ) , gt ( VAL_I { Int } , PREV_VAL { Int } , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom \forall { H { ArrayIntInt } , X { Int } , Y { Int } , F { SetInt } , PREV_VAL { Int } , MAX { Int } , .Patterns } \iff-lfp ( listSegmentLeftSorted ( H { ArrayIntInt } , X { Int } , Y { Int } , F { SetInt } , PREV_VAL { Int } , MAX { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( X { Int } , Y { Int } ) , \equals ( F { SetInt } , emptyset ) , \not ( gt ( PREV_VAL { Int } , MAX { Int } , .Patterns ) ) , .Patterns ) , \exists { F_I { SetInt } , VAL_I { Int } , X_I { Int } , .Patterns } \and ( listSegmentLeftSorted ( H { ArrayIntInt } , X_I { Int } , Y { Int } , F_I { SetInt } , VAL_I { Int } , MAX { Int } , .Patterns ) , gt ( X { Int } , 0 , .Patterns ) , \equals ( select ( H { ArrayIntInt } , X { Int } , .Patterns ) , X_I { Int } ) , \equals ( F { SetInt } , union ( F_I { SetInt } , singleton ( X { Int } , .Patterns ) , .Patterns ) ) , disjoint ( F_I { SetInt } , singleton ( X { Int } , .Patterns ) , .Patterns ) , \equals ( VAL_I { Int } , select ( H { ArrayIntInt } , plus ( X { Int } , 1 , .Patterns ) , .Patterns ) ) , gt ( VAL_I { Int } , PREV_VAL { Int } , .Patterns ) , \not ( gt ( VAL_I { Int } , MAX { Int } , .Patterns ) ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom functional ( listSegmentLeftSorted )
    </declaration> <declaration>
      axiom functional ( listSorted )
    </declaration> <declaration>
      symbol listSegmentLeftSorted ( ArrayIntInt , Int , Int , SetInt , Int , Int , .Sorts ) : Bool
    </declaration> <declaration>
      symbol listSorted ( ArrayIntInt , Int , SetInt , Int , .Sorts ) : Bool
    </declaration>
  </declarations>
</prover>
